File: 01/Exercise1/Program.cs
using System;
using System.IO;

namespace Exercise1
{
    class Program
    {
        static void Main(string[] args)
        {
            var expenses = File.ReadAllLines(args[0]);
            for (int i = 0; i < expenses.Length; i++)
            {
                var firstValue = int.Parse(expenses[i]);
                for (int j = i + 1; j < expenses.Length; j++)
                {
                    var secondValue = int.Parse(expenses[j]);
                    if (firstValue + secondValue == 2020)
                    {
                        Console.WriteLine(firstValue * secondValue);
                        Environment.Exit(0);
                    }
                }
            }
        }
    }
}

File: 01/fsharp/Exercise1/Program.fs
open System
open System.IO

let toPairs listToPair =
    let rec pairs pairingElement listToPair =
        let newPairs = listToPair |> List.map (fun element -> (pairingElement, element))
        match listToPair with
        | newPairingElement::remainingListToPair -> List.append newPairs (pairs newPairingElement remainingListToPair)
        | _ -> newPairs

    match listToPair with
    | pairingElement::listToPair -> pairs pairingElement listToPair
    | _ -> []

[<EntryPoint>]
let main argv =
    let entries =
        File.ReadAllLines(argv.[0])
        |> Array.map int
        |> Array.toList
    let answer = 
        entries
        |> toPairs
        |> List.find (fun (a, b) -> a + b = 2020)
        |> fun (a, b) -> a * b
    Console.WriteLine (answer)
    0 // return an integer exit code
